### PR TITLE
Alerting: Add lifecycle methods to the forked Alertmanager

### PR DESCRIPTION
**Pull request by santihernandezc**
This PR adds the following methods to the forked Alertmanager (two structs):
- `GetStatus()`
- `Cleanup()`
- `StopAndWait()`
- `Ready()`

### CODE REVIEW

File: pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go [hunk 1]
```diff
@@ -76,10 +76,20 @@ func (fam *RemotePrimaryForkedAlertmanager) TestTemplate(ctx context.Context, c
...
 
-func (fam *RemotePrimaryForkedAlertmanager) CleanUp() {}
+func (fam *RemotePrimaryForkedAlertmanager) CleanUp() {
+	// No cleanup to do in the remote Alertmanager.
+	fam.internal.CleanUp()
+}
 
-func (fam *RemotePrimaryForkedAlertmanager) StopAndWait() {}
+func (fam *RemotePrimaryForkedAlertmanager) StopAndWait() {
+	fam.internal.StopAndWait()
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: How does this work? Do these calls block? Is there no error that can be returned if it fails?

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: `StopAndWait()` stops the inhibitor, dispatcher, closes a few channels and waits using a waitgroup. It's blocking during that `waitgroup.Wait()` call, but this call to `internal.StopAndWait()` is no different from [the one we currently have](https://github.com/grafana/grafana/blob/68e37c3925080cf64a5e7570eabb042f19ca2dbf/pkg/services/ngalert/notifier/alertmanager.go#L158-L160). No values are returned.

Comment (grobinson-grafana) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: Right, but now we are communicating over a network when we do `fam.remote.StopAndWait()`. Do we need some kind of timeout via ctx?

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: No, we don't have anything to stop and wait on the remote Alertmanager side. The only thing we stop here are our senders (goroutines forwarding alerts to the remote Alertmanager).

Comment (grobinson-grafana) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: Ah I understand, thanks for clarification. I thought `fam.remote.StopAndWait()` did a network call.

Comment (grobinson-grafana) [inline pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: In that case why we would stop the internal sender? It shouldn't be running in remote primary mode, should it?

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: That's shutting down the entire internal Alertmanager, not the sender.
- `fam.internal.StopAndWait()` will shut down the internal Alertmanager's components (inhibitor, dispatcher, channels it needs in order to run).
- `fam.remote.StopAndWait()` will shut down the sender we have running for forwarding alerts to the remote Alertmanager.

We need the internal Alertmanager running even though we're not sending alerts there. We use the internal Alertmanager's `StopAndWait()` method to shut it down as a whole.

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go:85]: Both the dispatcher and the inhibitor are started in the internal Alertmanager's `ApplyConfig()` method, the code is in our [alerting repository](https://github.com/grafana/alerting/blob/bf12694896a843b7a77d7e7ef656013334214ef9/notify/grafana_alertmanager.go#L427-L437). We need to shut these down when exiting Grafana, and we can't start the internal Alertmanager without running these components.

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 2]
```diff
@@ -76,10 +76,20 @@ func (fam *RemoteSecondaryForkedAlertmanager) TestTemplate(ctx context.Context,
...
-func (fam *RemoteSecondaryForkedAlertmanager) StopAndWait() {}
+func (fam *RemoteSecondaryForkedAlertmanager) StopAndWait() {
+	fam.internal.StopAndWait()
+	fam.remote.StopAndWait()
+}
 
 func (fam *RemoteSecondaryForkedAlertmanager) Ready() bool {
-	return false
+	// Both Alertmanagers must be ready.
+	if ready := fam.remote.Ready(); !ready {
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:91]: You can make this simpler with:
```suggestion
	return fam.remote.Ready() && fam.internal.ready()
```

**Pull request merged**